param/__init__.py

```diff
@@ -1243,22 +1273,39 @@ def _validate(self, val):
             raise ValueError("%s not in Parameter %s's list of possible objects, "
                              "valid options include %s"%(val,attrib_name, items))
 
-    def _ensure_value_is_in_objects(self,val):
+    def _ensure_value_is_in_objects(self,val, name=None):
         """
         Make sure that the provided value is present on the objects list.
         Subclasses can override if they support multiple items on a list,
         to check each item instead.
         """
-        if not (val in self.objects):
-            self.objects.append(val)
+        if not (val in self.objects.values()):
+            self.objects[name_for_obj(val)]=val
 
     def get_range(self):
         """
         Return the possible objects to which this parameter could be set.
 
         (Returns the dictionary {object.name:object}.)
         """
-        return named_objs(self.objects, self.names)
+        return self.objects
+
+    @property
+    def names(self):
+        return self._names
+
+    @names.setter
+    def names(self, value):
+        self._names=value
+        self._objects=named_objs(self._objects, self._names)
+
+    @property
+    def objects(self):
+        return self._objects
+
+    @objects.setter
+    def objects(self, value):
+        self._objects=named_objs(value, self._names)
```

Comment:
When `objects` is updated, it would be nice to update the current value of this parameter automatically to be something from the object list, using code like `param_value = named_obj_default(self.objects, self.default)`. This would make GUI code simpler; e.g. if one parameter selects a `continent` that then determines the objects list of a `country` parameter, then it would be nice to update the actual value of the country as well. I don't know of any way to do that, though, given that the current value is typically in the dictionary of a Parameterized object, not on this Parameter object, but maybe this object can trace some pointers back to the parameterized in question?